000X-B-C-I-R-S-error-handling-codes.md

```diff
@@ -0,0 +1,307 @@
+  map<string, string> metadata = 3;
+}
+```
+
+### Error Status
+The properties of the **google.rpc.Status** will be populated as following:
+
+- **Code**  - Protocol level error code. These could be either gRPC or HTTP error codes. See (gRPC Codes ProtoBuf)[https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto]
+   
+   Example: "InvalidArgument Code = 3", "Internal Code = 13"
+
+- **Message**  - Error message.
+- **Details**  - A set of standard error payloads for error details. These list can be found in [Error Details](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto)
+   
+   Example: "ErrorInfo", "ResourceInfo"
+
+
+
+Below is partial table of the Standard Error code provided by gRPC and how they map to HTTP error codes. The entire list can found in the following links:
+- [Google API Error Handling]https://cloud.google.com/apis/design/errors#handling_errors
+- (gRPC Codes ProtoBuf)[https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto]
+
+
+|HTTP | gRPC      | Description |
+|---- | ----------- | ----------- |
+|200  |	OK		  | No error. |
+|400  |	INVALID_ARGUMENT |	Client specified an invalid argument. Check error message and error details for more information. |
+|400  |	FAILED_PRECONDITION |	Request can not be executed in the current system state, such as deleting a non-empty directory. |
+|400  |	OUT_OF_RANGE |	Client specified an invalid range. |
+|401  |	UNAUTHENTICATED	| Request not authenticated due to missing, invalid, or expired OAuth token. |
+|403  |	PERMISSION_DENIED |	Client does not have sufficient permission. This can happen because the OAuth token does not have the right scopes, the client doesn't have permission, or the API has not been enabled. |
+|404  |	NOT_FOUND |	A specified resource is not found. |
+|409  |	ABORTED	| Concurrency conflict, such as read-modify-write conflict. |
+
+
+### ErrorInfo
+The properties of the **type.googleapis.com/google.rpc.ErrorInfo** will be populated as following:
+
+- **Reason** - A combination of prefix from prefix of the table below plus the error condition code.
+    
+    Example: "DAPR_STATE_" + "ETAG_MISMATCH"
+
+- **Domain** - With the value `dapr.io`.
+
+- **Metadata** - A key/value map/dictionary data relevant to the error condition.
+
+### ResourceInfo
+The properties of the **type.googleapis.com/google.rpc.ResourceInfo** will be populated as following:
+
+- **ResourceType** - The building block type.
+   
+   Example: "StateStore"
+
+- **ResourceName** - The component name.
+   
+   Example: "REDIS"
+
+- **Owner**   - The owner of the component.
+
+- **Description** - Resource descrpition or error details.
+
+
+### Error Details Prefixes
+The following tables shows the propsosed error codes prefixes used in the **reason** for the **google.rpc.ErrorInfo** for various Dapr building blocks:
+
+
+**INIT**
+| Dapr Module | Prefix |
+| ----------- | ----------- |
+| CLI         | DAPR_CLI_INIT_* |
```

Comment:
Can we leave it at `DAPR_INIT_*`? We do have a separate dapr `CLI` and `daprd` is a binary and not a CLI.

Reply:
As we dig deeper into each building block we can redefine some of these.